hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/simple/HoodieGlobalSimpleIndex.java

```diff
@@ -113,42 +114,47 @@ protected List<Pair<String, HoodieBaseFile>> getAllBaseFilesInTable(
+        .map(entry -> {
           HoodieRecord<R> inputRecord = entry.getLeft();
           Option<Pair<String, HoodieRecordLocation>> partitionPathLocationPair = Option.ofNullable(entry.getRight().orElse(null));
-          List<HoodieRecord<R>> taggedRecords;
-
           if (partitionPathLocationPair.isPresent()) {
             String partitionPath = partitionPathLocationPair.get().getKey();
             HoodieRecordLocation location = partitionPathLocationPair.get().getRight();
-            if (config.getGlobalSimpleIndexUpdatePartitionPath() && !(inputRecord.getPartitionPath().equals(partitionPath))) {
-              // Create an empty record to delete the record in the old partition
-              HoodieRecord<R> deleteRecord = new HoodieAvroRecord(new HoodieKey(inputRecord.getRecordKey(), partitionPath), new EmptyHoodieRecordPayload());
-              deleteRecord.setCurrentLocation(location);
-              deleteRecord.seal();
-              // Tag the incoming record for inserting to the new partition
-              HoodieRecord<R> insertRecord = (HoodieRecord<R>) HoodieIndexUtils.getTaggedRecord(inputRecord, Option.empty());
-              taggedRecords = Arrays.asList(deleteRecord, insertRecord);
+            if (config.getGlobalSimpleIndexUpdatePartitionPath()) {
+              // The incoming record may need to be inserted to a new partition; keep the location info for merging later.
+              return Pair.of(inputRecord, partitionPathLocationPair);
             } else {
               // Ignore the incoming record's partition, regardless of whether it differs from its old partition or not.
               // When it differs, the record will still be updated at its old partition.
               HoodieRecord<R> newRecord = new HoodieAvroRecord(new HoodieKey(inputRecord.getRecordKey(), partitionPath), (HoodieRecordPayload) inputRecord.getData());
-              taggedRecords = Collections.singletonList((HoodieRecord<R>) HoodieIndexUtils.getTaggedRecord(newRecord, Option.ofNullable(location)));
+              return Pair.of(HoodieIndexUtils.getTaggedRecord(newRecord, Option.of(location)), Option.empty());
             }
           } else {
-            taggedRecords = Collections.singletonList((HoodieRecord<R>) HoodieIndexUtils.getTaggedRecord(inputRecord, Option.empty()));
+            return Pair.of(HoodieIndexUtils.getTaggedRecord(inputRecord, Option.empty()), Option.empty());
           }
-          return taggedRecords.iterator();
         });
+
+    if (config.getSimpleIndexUseCaching()) {
+      taggedRecordsAndLocationInfo.persist(new HoodieConfig(config.getProps())
+          .getString(HoodieIndexConfig.SIMPLE_INDEX_INPUT_STORAGE_LEVEL_VALUE));
+    }
+    HoodieData<HoodieRecord<R>> finalTaggedRecords = mergeForPartitionUpdates(taggedRecordsAndLocationInfo, config, hoodieTable);
```

Comment:
if update partition path config is set to false, I thought we will avoid this new code path (mergeForPartitionUpdates) ?

Reply:
good catch. fixed!